### DISCUSSION TITLE
Can you add IHaskell?

### DISCUSSION THREAD

**Discussion by FedericoMontana** [srid/haskell-template#15] [General]
Was wondering if it could be possible to add IHaskell as part of the package

**Comment by srid** [comment]
You can add it to the [cabal file](https://github.com/srid/haskell-template/blob/c94f12580fee8f6ccac58cccf8261cb86eb0af2a/haskell-template.cabal#L26-L41) in your project.

It shouldn't be in this template because IHaskell is typically not used in a random Haskell project, unlike say `base` or `containers.`

**Reply by FedericoMontana** [reply]
Thank you. Just wanted to show some appreciation as I just started with Haskell and spent days with errors with my M1. Finally got HLS working with your template. Didn't yet give up thanks to this.

If I add ", IHaskell   " entry in the cabal file in theory it will install the required configuration and out of the box work? If you happen to have the time to let me know the steps, I'd highly appreciate it.

**Reply by srid** [reply]
Great to hear!

> If I add ", IHaskell " entry in the cabal file in theory it will install the required configuration and out of the box work? 

Exactly. See https://github.com/srid/ihaskell-demo/commit/1720c2dfa71fa7565eeebe8efb02c5b726a848cf

This will make IHaskell Haskell library available, but if you are referring to the executable that's a different matter. For that, run:

```sh
nix run nixpkgs#haskellPackages.ihaskell
```

If you are on NixOS, there is also [`services.ihaskell`](https://search.nixos.org/options?channel=21.11&show=services.ihaskell.enable&from=0&size=50&sort=relevance&type=packages&query=ihaskell) which runs the IHaskell notebook as a service.

**Reply by FedericoMontana** [reply]
Thanks. Last question I promise. I will add that library, and then how would I reload the project so that it takes effect? And the nix run command should be run within my project right? 
Thanks again Srid!

**Reply by srid** [reply]
> I will add that library, and then how would I reload the project so that it takes effect?

Re-run `nix develop` (or reload VSCode). As well as re-run `bin/run`.

> And the nix run command should be run within my project right?

It can be run from anywhere.

But if you add `haskellPackages.ihaskell` to [`shellDeps`](https://github.com/srid/haskell-template/blob/c94f12580fee8f6ccac58cccf8261cb86eb0af2a/flake.nix#L34-L42), `nix develop` and then `ihaskell` will also work.

**Reply by FedericoMontana** [reply]
Thank you! did all that. But when I create a foo.ipynb within my project created with your template, with the haskellpackages.ihaskell added, how do I connect it to a kernel for it to work?

**Reply by srid** [reply]
I'm not very familiar with IHaskell in its current state, but you might have better luck discussing this in the [Nix forums](https://discourse.nixos.org/).
